pkg/apis/traceflow/v1alpha1/types.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Phase string
+
+const (
+	Initial   Phase = "Pending"
+	Running   Phase = "Running"
+	Succeeded Phase = "Succeeded"
+	Failed    Phase = "Failed"
+)
+
+type Component string
+
+const (
+	SpoofGuard    Component = "SpoofGuard"
+	LB            Component = "LB"
+	Routing       Component = "Routing"
+	NetworkPolicy Component = "NetworkPolicy"
+	Forwarding    Component = "Forwarding"
+)
+
+type Action string
+
+const (
+	Delivered Action = "Delivered"
+	Received  Action = "Received"
+	Forwarded Action = "Forwarded"
+	Dropped   Action = "Dropped"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Traceflow struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   `json:"spec,omitempty"`
+	Status `json:"status,omitempty"`
+	Reason *string `json:"reason,omitempty"`
```

Comment:
"Reason" should be in "Status" and normally not a pointer: https://github.com/kubernetes/kubernetes/blob/15e95e48964d4e800f7210a249f9122ee7d79fcc/pkg/apis/core/types.go#L3068.

Reply:
Thanks for the example. Updated. However, I made it this way because of another example you provided. There `Reason` is outside `Status` and is a pointer. Maybe I misunderstood it.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties